beluga/include/beluga/algorithm/amcl_core.hpp

```diff
@@ -0,0 +1,240 @@
+  double alpha_fast = 0.1;
+  /// Used as part of the kld sampling mechanism.
+  double kld_epsilon = 0.05;
+  /// Used as part of the kld sampling mechanism.
+  double kld_z = 3.0;
+};
+
+/// Implementation of the Adaptive Monte Carlo Localization (AMCL) algorithm.
+/**
+ * \tparam MotionModel Class representing a motion model. Must satisfy \ref MotionModelPage.
+ * \tparam SensorModel Class representing a sensor model. Must satisfy \ref SensorModelPage.
+ * \tparam RandomStateGenerator A callable able to produce random states, optionally based on the current particles
+ * state.
+ * Class 'T' is a valid RandomStateGenerator if given 't' a possibly const instance of 'T' any of the following
+ * conditions are true:
+ * - t can be called without arguments returning an instance of 'SensorModel::state_type'
+ *   representing a random state.
+ * - t can be called with `const beluga::TupleVector<ParticleType>&> returning a callable
+ *   that can be called without arguments and return an instance of 'SensorModel::state_type'.
+ * \tparam WeightT Type to represent a weight of a particle.
+ * \tparam ParticleType Full particle type, containing state, weight and possibly
+ * other information .
+ * \tparam ExecutionPolicy Execution policy for particles processing.
+ */
+template <
+    class MotionModel,
+    class SensorModel,
+    class RandomStateGenerator,
+    typename WeightT = beluga::Weight,
+    class ParticleType = std::tuple<typename SensorModel::state_type, WeightT>,
+    class ExecutionPolicy = std::execution::sequenced_policy>
+
+class Amcl {
+  static_assert(
+      std::is_same_v<ExecutionPolicy, std::execution::parallel_policy> or
+      std::is_same_v<ExecutionPolicy, std::execution::sequenced_policy>);
+
+  using particle_type = ParticleType;
+  using measurement_type = typename SensorModel::measurement_type;
+  using state_type = typename SensorModel::state_type;
+  using map_type = typename SensorModel::map_type;
+  using spatial_hasher_type = spatial_hash<state_type>;
+  using random_state_generator_type = RandomStateGenerator;
+  using estimation_type =
+      std::invoke_result_t<decltype(beluga::estimate<std::vector<state_type>>), std::vector<state_type>>;
+
+ public:
+  /// Construct a AMCL instance.
+  /**
+   * \param motion_model Motion model instance.
+   * \param sensor_model Sensor model Instance.
+   * \param random_state_generator A callable able to produce random states, optionally based on the current particles
+   * state.
+   * \param spatial_hasher A spatial hasher instance capable of computing a hash out of a particle state.
+   * \param params Parameters for AMCL implementation.
+   * \param execution_policy Policy to use when processing particles.
+   */
+  Amcl(
+      MotionModel motion_model,
+      SensorModel sensor_model,
+      RandomStateGenerator random_state_generator,
+      spatial_hasher_type spatial_hasher,
```

Comment:
@serraramiro1 meta: we need to sort out the mixed casing we have for types here and there.

Reply:
Yeah, our `snake_case` class names is all over the place.
I'll leave this as is for now if that's OK